// popmotion-pose/lib/flip.js
'use strict';

exports.__esModule = true;
exports.flipPose = exports.isFlipPose = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _value = require('popmotion/reactions/value');

var _value2 = _interopRequireDefault(_value);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var positionalProps = new Set(['width', 'height', 'top', 'left', 'bottom', 'right']);
var checkPositionalProp = function checkPositionalProp(key) {
  return positionalProps.has(key);
};
var hasPositionalProps = function hasPositionalProps(pose) {
  return Object.keys(pose).some(checkPositionalProp);
};
var isFlipPose = exports.isFlipPose = function isFlipPose(pose, key) {
  return hasPositionalProps(pose) || key === 'flip';
};

var setValue = function setValue(_ref, key, to) {
  var values = _ref.values,
      elementStyler = _ref.elementStyler;

  if (values.has(key)) {
    // Here, if we already have the value, we update it twice.
    // Because of stylefire's render batching, this isn't going
    // to actually render twice, but because we're making
    // the value jump a great distance, we want to reset the velocity
    // to 0, rather than something arbitrarily high
    var val = values.get(key).value;
    val.update(to);
    val.update(to);
  } else {
    values.set(key, { value: (0, _value2.default)(to, function (v) {
        return elementStyler.set(key, v);
      }) });
  }
};

var explicitlyFlipPose = function explicitlyFlipPose(state, nextPose) {
  var dimensions = state.dimensions,
      elementStyler = state.elementStyler;

  dimensions.measure();

  var width = nextPose.width,
      height = nextPose.height,
      top = nextPose.top,
      left = nextPose.left,
      bottom = nextPose.bottom,
      right = nextPose.right,
      remainingPose = _objectWithoutProperties(nextPose, ['width', 'height', 'top', 'left', 'bottom', 'right']);

  elementStyler.set({ width: width, height: height, top: top, left: left, bottom: bottom, right: right }).render();

  return implicitlyFlipPose(state, remainingPose);
};

var implicitlyFlipPose = function implicitlyFlipPose(state, nextPose) {
  var dimensions = state.dimensions,
      element = state.element,
      elementStyler = state.elementStyler;

  if (!dimensions.has()) return;

  var flipPose = {};
  var prev = dimensions.get();
  var next = element.getBoundingClientRect();

  if (prev.width !== next.width) {
    setValue(state, 'scaleX', prev.width / next.width);
    flipPose.scaleX = 1;
  }

  if (prev.height !== next.height) {
    setValue(state, 'scaleY', prev.height / next.height);
    flipPose.scaleY = 1;
  }

  var originX = prev.top === next.top ? 0 : prev.bottom === next.bottom ? '100%' : '50%';

  var originY = prev.left === next.left ? 0 : prev.right === next.right ? '100%' : '50%';

  // Here, if we already have the value, we update it twice.
  // Because of stylefire's render batching, this isn't going
  // to actually render twice, but because we're making
  // the value jump a great distance, we want to reset the velocity
  // to 0, rather than something arbitrarily high

  elementStyler.set({ originX: originX, originY: originY }).render();

  return _extends({}, nextPose, flipPose);
};

var flipPose = exports.flipPose = function flipPose(state, nextPose) {
  return hasPositionalProps(nextPose) ? explicitlyFlipPose(state, nextPose) : implicitlyFlipPose(state, nextPose);
};
