// react-pose/lib/PoseElement.js
'use strict';

exports.__esModule = true;
exports.PoseParentContext = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _utils = require('./utils');

var _popmotionPose = require('popmotion-pose');

var _popmotionPose2 = _interopRequireDefault(_popmotionPose);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PoseParentContext = exports.PoseParentContext = _react2.default.createContext({});

var PoseElement = function (_React$PureComponent) {
  _inherits(PoseElement, _React$PureComponent);

  function PoseElement(props) {
    _classCallCheck(this, PoseElement);

    var _this = _possibleConstructorReturn(this, _React$PureComponent.call(this, props));

    _this.setRef = function (ref) {
      return _this.ref = ref;
    };

    _this.children = new Set();
    _this.childrenHandlers = {
      onMount: function onMount(child) {
        return _this.children.add(child);
      },
      onUnmount: function onUnmount(child) {
        _this.children.remove(child);
        _this.poser.removeChild(child);
      },
      getCurrentFromParent: function getCurrentFromParent() {
        return _this.props.current;
      }

      /**
       * Prop selectors
       * =============================================
       */
    };
    var poses = props.poses,
        current = props.current,
        getCurrentFromParent = props.getCurrentFromParent;

    var currentPose = _this.getInitialPose();
    _this.style = poses && currentPose ? (0, _utils.posesToStyles)(poses[currentPose]) : {};
    return _this;
  }

  /**
   * Lifecycle
   * =============================================
   */


  PoseElement.prototype.componentDidMount = function componentDidMount() {
    var _this2 = this;

    if (!this.ref) return;

    this.poser = (0, _popmotionPose2.default)(this.ref, this.getPoseProps());
    var onMount = this.props.onMount;

    if (onMount) onMount(this.poser);
    if (this.children.size) this.children.forEach(function (child) {
      return _this2.poser.addChild(child);
    });
  };

  PoseElement.prototype.componentDidUpdate = function componentDidUpdate(_ref) {
    var prevCurrent = _ref.current;
    var current = this.props.current;

    if (prevCurrent !== current) this.poser.set(current);
  };

  PoseElement.prototype.componentWillUnmount = function componentWillUnmount() {
    if (!this.poser) return;

    var onUnmount = this.props.onUnmount;

    if (onUnmount) onUnmount(this.poser);
    this.poser.destroy();
  };

  /**
   * Children handlers
   * =============================================
   */


  PoseElement.prototype.getPoseProps = function getPoseProps() {
    var _props = this.props,
        elementType = _props.elementType,
        children = _props.children,
        draggable = _props.draggable,
        poses = _props.poses;


    return _extends({
      draggable: draggable,
      initialPose: this.getInitialPose()
    }, poses);
  };

  PoseElement.prototype.getDomProps = function getDomProps(_ref2) {
    var draggable = _ref2.draggable,
        poses = _ref2.poses,
        current = _ref2.current,
        style = _ref2.style,
        onMount = _ref2.onMount,
        onUnmount = _ref2.onUnmount,
        getCurrentFromParent = _ref2.getCurrentFromParent,
        props = _objectWithoutProperties(_ref2, ['draggable', 'poses', 'current', 'style', 'onMount', 'onUnmount', 'getCurrentFromParent']);

    props.style = _extends({}, style, this.style);
    props.ref = this.setRef;
    return props;
  };

  PoseElement.prototype.getInitialPose = function getInitialPose() {
    var _props2 = this.props,
        current = _props2.current,
        getCurrentFromParent = _props2.getCurrentFromParent;

    return current ? current : getCurrentFromParent ? getCurrentFromParent() : undefined;
  };

  PoseElement.prototype.render = function render() {
    var _props3 = this.props,
        elementType = _props3.elementType,
        children = _props3.children,
        remaining = _objectWithoutProperties(_props3, ['elementType', 'children']);

    return _react2.default.createElement(
      PoseParentContext.Provider,
      { value: this.childrenHandlers },
      _react2.default.createElement.apply(_react2.default, [elementType, this.getDomProps(remaining)].concat(children))
    );
  };

  return PoseElement;
}(_react2.default.PureComponent);

exports.default = PoseElement;
