// react-pose/lib/utils.js
'use strict';

exports.__esModule = true;
exports.posesToStyles = undefined;

var _stylefire = require('stylefire');

var _popmotion = require('popmotion');

var pipe = _popmotion.transform.pipe;

// Consider mapping to stylefire

var cssAliasMap = {
  x: 'translateX',
  y: 'translateY',
  z: 'translateZ',
  originX: 'transformOriginX',
  originY: 'transformOriginY'
};

var nextCharToUppercase = function nextCharToUppercase(m) {
  return m[1].toUpperCase();
};
var dashToCamel = function dashToCamel(str) {
  return str.replace(/-([a-z])/g, nextCharToUppercase);
};

var addRule = function addRule(map, rule) {
  var _rule$split = rule.split(':'),
      key = _rule$split[0],
      value = _rule$split[1];

  map[dashToCamel(key)] = value.trim();
  return map;
};

var cssStringToObject = function cssStringToObject(css) {
  return css.split(';').filter(Boolean).reduce(addRule, {});
};

var mapAliased = function mapAliased(css) {
  var aliased = {};
  var keys = Object.keys(css);
  var numKeys = keys.length;
  for (var i = 0; i < numKeys; i++) {
    var key = keys[i];
    var aliasedKey = cssAliasMap[key] ? cssAliasMap[key] : key;
    aliased[aliasedKey] = css[key];
  }
  return aliased;
};

var posesToStyles = exports.posesToStyles = pipe(mapAliased, _stylefire.buildStyles, cssStringToObject);
